examples/localstack/localstack.go

```diff
@@ -0,0 +1,190 @@
+package localstack
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/docker/go-connections/nat"
+	"github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/internal/testcontainersdocker"
+	"github.com/testcontainers/testcontainers-go/wait"
+	"golang.org/x/mod/semver"
+)
+
+const defaultPort = 4566
+const defaultRegion = "us-east-1"
+const defaultVersion = "0.11.2"
+const hostnameExternalEnvVar = "HOSTNAME_EXTERNAL"
+
+const defaultAccessKeyID = "accesskey"
+const defaultSecretAccessKey = "secretkey"
+const defaultToken = "token"
+
+// LocalStackContainer represents the LocalStack container type used in the module
+type LocalStackContainer struct {
+	testcontainers.Container
+	Credentials     Credentials
+	Region          string
+	EnabledServices map[string]Service
+}
+
+type Credentials struct {
+	AccessKeyID     string
+	SecretAccessKey string
+	Token           string
+}
+
+// ServicePort returns the port of the given service
+func (l *LocalStackContainer) ServicePort(ctx context.Context, service EnabledService) (nat.Port, error) {
+	if _, ok := l.EnabledServices[service.Name()]; !ok {
+		return "", fmt.Errorf("service %s is not enabled", service.Name())
+	}
+
+	internalPort := l.EnabledServices[service.Name()].servicePort()
+
+	p, err := nat.NewPort("tcp", fmt.Sprintf("%d", internalPort))
+	if err != nil {
+		return "", err
+	}
+
+	return l.MappedPort(ctx, p)
+}
+
+func runInLegacyMode(version string) bool {
+	if version == "latest" {
+		return false
+	}
+
+	if !strings.HasPrefix(version, "v") {
+		version = fmt.Sprintf("v%s", version)
+	}
+
+	if semver.IsValid(version) {
+		return semver.Compare(version, "v0.11") < 0 // version < v0.11
+	}
+
+	fmt.Printf("Version %s is not a semantic version, LocalStack will run in legacy mode.\n", version)
+	fmt.Printf("Consider using \"StartContainer(ctx, NoopOverrideContainerRequest, WithLegacyMode)\" constructor if you want to enable legacy mode.")
+	return true
+}
+
+// StartContainer creates an instance of the LocalStack container type
+func StartContainer(ctx context.Context, overrideReq overrideContainerRequestOption, opts ...localStackContainerOption) (*LocalStackContainer, error) {
+	credentials := Credentials{
+		AccessKeyID:     defaultAccessKeyID,
+		SecretAccessKey: defaultSecretAccessKey,
+		Token:           defaultToken,
+	}
+
+	req := testcontainers.ContainerRequest{
+		Image:      "localstack/localstack",
```

Comment:
what if there is a custom image? As part of the example this is good but thinking about modules this will not allow to use custom images, not only those hosted in a registry but what if my image is called `my-localstack`.

Also, I don't see a test which can override this image with, for example, `localstack/localstack:0.13.0`

Reply:
Because it's possible to pass a `OverrideContainerRequest` optional func, where you can customise as much as you need from the original container request, I did not add a explicit optional func to change it. Please take a look at the unit tests for that func type:

```go
func TestWithContainerRequest(t *testing.T) {
	req := testcontainers.ContainerRequest{
		Env:          map[string]string{},
		Image:        "foo",
		ExposedPorts: []string{},
		SkipReaper:   false,
		WaitingFor: wait.ForNop(
			func(ctx context.Context, target wait.StrategyTarget) error {
				return nil
			},
		),
		Networks: []string{"foo", "bar", "baaz"},
		NetworkAliases: map[string][]string{
			"foo": {"foo0", "foo1", "foo2", "foo3"},
		},
	}

	merged := OverrideContainerRequest(testcontainers.ContainerRequest{
		Env: map[string]string{
			"FOO": "BAR",
		},
		Image:    "bar",
		Networks: []string{"foo1", "bar1"},
		NetworkAliases: map[string][]string{
			"foo1": {"bar"},
		},
		WaitingFor: wait.ForLog("foo"),
		SkipReaper: true,
	})(req)

	assert.Equal(t, "BAR", merged.Env["FOO"])
	assert.True(t, merged.SkipReaper)
	assert.Equal(t, "bar", merged.Image)
	assert.Equal(t, []string{"foo1", "bar1"}, merged.Networks)
	assert.Equal(t, []string{"foo0", "foo1", "foo2", "foo3"}, merged.NetworkAliases["foo"])
	assert.Equal(t, []string{"bar"}, merged.NetworkAliases["foo1"])
	assert.Equal(t, wait.ForLog("foo"), merged.WaitingFor)
}
```

Therefore, because passing the `OverrideContainerRequest` functional opt to the `StartContainer` function, in which the functional opt receive a way to change the image, then we are good to go.

A different story would be that we want to explicitly make certain functions public, like `WithImage`.